sdk/core/azure-core/inc/azure/core/http/websockets/websockets_transport.hpp

```diff
@@ -0,0 +1,252 @@
+     *
+     * @returns true if the transport has native websocket support, false otherwise.
+     */
+    virtual bool HasNativeWebsocketSupport() = 0;
+
+    /**************/
+    /* Native WebSocket support functions*/
+    /**************/
+    /**
+     * @brief Gracefully closes the WebSocket, notifying the remote node of the close reason.
+     *
+     * @param status Status value to be sent to the remote node. Application defined.
+     * @param disconnectReason UTF-8 encoded reason for the disconnection. Optional.
+     * @param context Context for the operation.
+     */
+    virtual void CloseSocket(
+        uint16_t status,
+        std::string const& disconnectReason,
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**
+     * @brief Closes the WebSocket.
+     *
+     * Does not notify the remote endpoint that the socket is being closed.
+     *
+     */
+    virtual void Close() = 0;
+
+    /**
+     * @brief Retrieve the information associated with a WebSocket close response.
+     *
+     * @param context Context for the operation.
+     *
+     * @returns a tuple containing the status code and string.
+     */
+    virtual std::pair<uint16_t, std::string> GetCloseSocketInformation(
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**
+     * @brief Send a frame of data to the remote node.
+     *
+     * @param frameType Frame type sent to the server, Text or Binary.
+     * @param frameData Frame data to be sent to the server.
+     * @param context Context for the operation.
+     */
+    virtual void SendFrame(
+        WebSocketFrameType frameType,
+        std::vector<uint8_t> const& frameData,
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**
+     * @brief Receive a frame from the remote WebSocket server.
+     *
+     * @param context Context for the operation.
+     *
+     * @returns a tuple containing the Frame data received from the remote server and the type of
+     * data returned from the remote endpoint
+     */
+    virtual std::pair<WebSocketFrameType, std::vector<uint8_t>> ReceiveFrame(
+        Azure::Core::Context const& context)
+        = 0;
+
+    /**************/
+    /* Non Native WebSocket support functions */
+    /**************/
+
+    // Implement a buffered stream reader
+    class BufferedStreamReader {
+      std::shared_ptr<Azure::Core::Http::WebSockets::WebSocketTransport> m_transport;
+      std::unique_ptr<Azure::Core::IO::BodyStream> m_initialBodyStream;
+      constexpr static size_t m_bufferSize = 1024;
+      uint8_t m_buffer[m_bufferSize]{};
+      size_t m_bufferPos = 0;
+      size_t m_bufferLen = 0;
+      bool m_eof = false;
+
+    public:
+      explicit BufferedStreamReader() = default;
+      ~BufferedStreamReader() = default;
+
+      void SetInitialStream(std::unique_ptr<Azure::Core::IO::BodyStream>& stream)
+      {
+        m_initialBodyStream = std::move(stream);
```

Comment:
The initial stream cannot be a constructor argument because it's not created until the initial HTTP request, which happens after the ctor.

This method exists because there is at least one WebSocket implementation that sends the first frame of data in the body of the HTTP response and I need to "prime the pump" with that first frame of data.

Reply:
FWIW, this was an experiment that accidentally got checked in, I removed it.